### PR TITLE
Make custom REPL commands accept an argument

### LINKED ISSUE #1008: Built-in tap current form.
This ticket is for designing/approving a built-in tap current form `calva.tapCurrentForm`.
`Alt+Shift+Enter` seems to be free, I propose it as keybinding.

Should we have tap top level form?
Related: https://github.com/seancorfield/vscode-clover-setup/blob/develop/keybindings.json

### LINKED ISSUE #1011: Add args to Custom REPL Commands
I want the **Run Custom REPL command** feature to work like described below:

Related to #1008 

# Custom REPL Commands

Calva supports configuration of custom command snippets that you can evaluate in the REPL at will. If your workflow has you repeatedly evaluate a particular piece of code, you can use the setting `calva.customREPLCommandSnippets` to configure it. Then either bind keyboard shortcuts to them or use the command **Run Custom REPL Command** to access it. The command will give you a menu with the snippets you have configured.

The `calva.customREPLCommandSnippets` is an array of objects with the following fields (required fields in **bold**):

* **`name`**: The name of the snippet as it will appear in the picker menu
* **`snippet`**: The code that will be evaluated
* `key`: A key can be used to reference the snippet from **Run Custom REPL Command** keyboard shortcut arguments. It will also be used in the quick-pick menu.
* `ns`: A namespace to evaluate the command in. If omitted the command will be executed in the namespace of the current editor.
* `repl`: Which repl session to use for the evaluation. Either `"clj"` or `"cljs"`. Omit if you want to use the session of the current editor.

There are also substitutions available, which will take elements from the current state of Calva and splice them in to the text of your command before executing it. They are

* `$line`: Current line number in editor
* `$column`: Current column number in editor
* `$file`: Full name of current file edited
* `ns`: The namespace used for evaluating the command
* `$current-form`: The text of the [current form](eval-tips.md#current-form)
* `$top-level-form` The text of the [current top level form](eval-tips.md#current-top-level-form)
* `$current-fn`: The sexpr/form at call position in the current list, e.g. `str` with `(defn foo [] (str "foo" "bar|"))`
* `$top-level-defined-symbol`: The second symbol of the top level form, e.g. `foo` with `(defn foo [] (str "foo" "bar|"))`

Consider these settings:

```json
        "calva.customREPLCommandSnippets": [
        {
            "name": "Foo",
            "snippet": "(println :foo)",
            "ns": "acme.test.foo-test",
            "repl": "cljs"
        },
        {
            "name": "Bar",
            "snippet": "(println :bar $line)",
            "ns": "acme.test.bar-test",
            "repl": "clj"
        },
        {
            "name": "Refresh",
            "key": "r",
            "snippet": "(refresh)",
            "repl": "clj"
        },
        {
            "name": "Call Current Form",
            "repl": "clj",
            "snippet": "($current-form)"
        },
        {
            "name": "Call Current Top Level Form",
            "key": "call-t",
            "repl": "clj",
            "snippet": "($top-level-form)"
        },
        {
            // You don't need to configure this one,
            // there is a built-in command for tapping the current form
            "name": "Tap Current Form",
            "key": "tap",
            "repl": "clj",
            "snippet": "(tap> $current-form)"
        },
        {
            "name": "Evaluate Current Function Symbol",
            "repl": "clj",
            "snippet": "$current-fn"
        },
        {
            "name": "Call Top Level Defined Symbol",
            "repl": "clj",
            "snippet": "($top-level-defined-symbol)"
        }
    ]
```

Issuing **Run Custom REPL Command** will then render this VS Code menu:

image

The items are numbered for you so that you can choose them in predictable way. The default keyboard shortcut for the command is `ctrl+alt+space`. Which means that to execute the **Call Current Form** custom command, you could do:

`ctrl+alt+space 4 enter`.

## Binding keyboard shortcuts

Some custom REPL commands might be so central to your workflow that you want to bind keyboard shortcuts to them directly. There are two ways to do this:

1. Bind `calva.runCustomREPLCommand` to a shortcut with whatever code you want to evaluate in `args` key. You have access to the substitution variables here as well.
2. Bind `calva.runCustomREPLCommand` to a shortcut referencing the `key` of on of your `calva.customREPLCommandSnippets`.

Given the above example settings, here are two shortcut definitions that both `tap>` the current form:

```json
    {
        "key": "ctrl+alt+shift+t t",
        "command": "calva.runCustomREPLCommand",
        "args": "(tap> $current-form)"
    },
    {
        "key": "ctrl+alt+shift+t t",
        "command": "calva.runCustomREPLCommand",
        "args": "tap"
    }
```

(Again, none of them is needed, the built in command for tapping the current form is default bound to the shortcut `ctrl+shift+t t`.)

### PR DESCRIPTION
**Pull request by PEZ**
## What has Changed?

See #1011 for what this enables.

Fixes #1011
Fixes #1008

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [ ] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test. NB: *There is a CircleCI bug that makes the Artifacts hard to find. Please see [this issue](https://discuss.circleci.com/t/artifacts-tab-not-showing-unless-logged-in/32433) for workarounds.*
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Added to or updated docs in this branch, if appropriate

## The Calva Team PR Checklist:

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and tested it there if so.
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse, @bpringe

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -164,10 +164,9 @@ async function activate(context: vscode.ExtensionContext) {
     context.subscriptions.push(vscode.commands.registerCommand('calva.refresh', refresh.refresh));
     context.subscriptions.push(vscode.commands.registerCommand('calva.refreshAll', refresh.refreshAll));
     context.subscriptions.push(vscode.commands.registerCommand('calva.debug.instrument', eval.instrumentTopLevelForm));
-    context.subscriptions.push(vscode.commands.registerCommand('calva.runCustomREPLCommand', async () => {
-        await snippets.evaluateCustomCommandSnippetCommand();
-        outputWindow.appendPrompt();
-    }));
+    context.subscriptions.push(vscode.commands.registerCommand('calva.runCustomREPLCommand', snippets.evaluateCustomCodeSnippetCommand));
+    context.subscriptions.push(vscode.commands.registerCommand('calva.tapSelection', () => snippets.evaluateCustomCodeSnippetCommand("(tap> $current-form)")));
+    context.subscriptions.push(vscode.commands.registerCommand('calva.tapCurrentTopLevelForm', () => snippets.evaluateCustomCodeSnippetCommand("(tap> $top-level-form)")));
```
Comment (bpringe) [inline src/extension.ts:169]: I love good composition like this :smiley:

**Review by bpringe — APPROVED** [review]

**Review by bpringe — APPROVED** [review]

**Comment by bpringe** [comment]
Accidentally self-requested a review. Now it's doubly approved.

**Comment by bpringe** [comment]
I need to retract that approval :sweat_smile: 

With the following custom snippets, I get errors when running them while the cursor is in the repl window. I was also getting errors when the cursor was in a .clj file but that stopped happening and I'm not certain why.

```json
"calva.customREPLCommandSnippets": [
        {
            "name": "Foo",
            "snippet": "(println :foo)",
            "repl": "clj"
        },
        {
            "name": "Bar",
            "snippet": "(println :bar $line)",
            "repl": "clj"
        }
    ]
 ```
 
 1. Connect repl
 2. Load a file (this probably shouldn't be necessary and I don't know if it is, just doing to show it still errors)
 3. Run the Foo snippet _with the cursor in the repl window_, get the following error (it works if run from an editor).
4. Run the Bar snippet _with the cursor in the repl window_, get the following error (it works if run from an editor).

Both 3 and 4 result in the same error:

image

File: src/extension.ts (hunk 1 above)
Comment (PEZ) [reply src/extension.ts:169]: Haha, it's a good sign. 😄

**Pull request merged**